File: android/Showcase/src/main/java/com/spruceid/mobilesdkexample/wallet/HacApplicationListItem.kt
package com.spruceid.mobilesdkexample.wallet

import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.spruceid.mobile.sdk.CredentialStatusList
import com.spruceid.mobilesdkexample.credentials.CredentialStatusSmall
import com.spruceid.mobilesdkexample.db.HacApplications
import com.spruceid.mobilesdkexample.ui.theme.ColorBase300
import com.spruceid.mobilesdkexample.ui.theme.ColorStone950
import com.spruceid.mobilesdkexample.ui.theme.Inter
import com.spruceid.mobilesdkexample.viewmodels.HacApplicationsViewModel

@Composable
fun HacApplicationListItem(
    application: HacApplications?,
    startIssuance: (String, suspend () -> Unit) -> Unit,
    hacApplicationsViewModel: HacApplicationsViewModel?
) {
    var credentialOfferUrl by remember { mutableStateOf<String?>(null) }
    var credentialStatus by remember { mutableStateOf(CredentialStatusList.UNDEFINED) }

    LaunchedEffect(application) {
        if (application != null) {
            try {
                val status = hacApplicationsViewModel?.issuanceClient?.checkStatus(
                    issuanceId = application.issuanceId,
                    walletAttestation = hacApplicationsViewModel.getWalletAttestation()!!
                )
                if (status?.state == "ReadyToProvision") {
                    credentialStatus = CredentialStatusList.READY
                }
                credentialOfferUrl = status?.openidCredentialOffer
            } catch (e: Exception) {
                println(e.message)
            }
        } else {
            credentialStatus = CredentialStatusList.PENDING
        }
    }

    Column(
        Modifier
            .fillMaxWidth()
            .padding(vertical = 10.dp)
            .border(
                width = 1.dp,
                color = ColorBase300,
                shape = RoundedCornerShape(8.dp)
            )
            .padding(12.dp)
            .clickable(
                enabled = credentialOfferUrl != null,
                onClick = {
                    credentialOfferUrl?.let { url ->
                        startIssuance(url) {
                            application?.let {
                                hacApplicationsViewModel?.deleteApplication(application.id)
                            }
                        }
                    }
                }
            )
    ) {
        Column {
            Text(
                text = "Mobile Drivers License",
                fontFamily = Inter,
                fontWeight = FontWeight.Medium,
                fontSize = 20.sp,
                color = ColorStone950,
                modifier = Modifier.padding(bottom = 8.dp)
            )
            CredentialStatusSmall(status = credentialStatus)
        }
        Spacer(modifier = Modifier.weight(1f))
    }
}

File: ios/Showcase/Targets/AppUIKit/Sources/wallet/HacApplicationListItem.swift
import SpruceIDMobileSdk
import SpruceIDMobileSdkRs
import SwiftUI

struct HacApplicationListItem: View {
    @EnvironmentObject var hacApplicationObservable: HacApplicationObservable

    let application: HacApplication?
    let startIssuance: (String) -> Void
    @State var credentialOfferUrl: String?
    @State var credentialStatus: CredentialStatusList = .undefined

    var body: some View {
        HStack {
            VStack(alignment: .leading) {
                VStack(alignment: .leading, spacing: 12) {
                    Text("Mobile Drivers License")
                        .font(
                            .customFont(
                                font: .inter,
                                style: .semiBold,
                                size: .h1
                            )
                        )
                        .foregroundStyle(Color("ColorStone950"))
                    CredentialStatusSmall(
                        status: credentialStatus
                    )
                }
                .padding(.leading, 12)
            }
            Spacer()
        }
        .padding(12)
        .background(
            RoundedRectangle(cornerRadius: 8)
                .fill(Color.white)
                .shadow(color: .black.opacity(0.03), radius: 5)
        )
        .overlay(
            RoundedRectangle(cornerRadius: 8)
                .stroke(Color("ColorBase300"), lineWidth: 1)
        )
        .padding(12)
        .onTapGesture {
            if let credentialOfferUrl = credentialOfferUrl {
                startIssuance(credentialOfferUrl)
            }
        }
        .onAppear {
            if let application = application {
                Task {
                    do {
                        let walletAttestation =
                            try await hacApplicationObservable
                            .getWalletAttestation()
                            .unwrap()

                        let status =
                            try await hacApplicationObservable.issuanceClient
                            .checkStatus(
                                issuanceId: application.issuanceId,
                                walletAttestation: walletAttestation
                            )
                        if status.state == "ReadyToProvision" {
                            credentialStatus = .ready
                        }
                        credentialOfferUrl = status.openidCredentialOffer
                    } catch {
                        print(error.localizedDescription)
                    }
                }
            } else {
                credentialStatus = .pending
            }
        }
    }
}
